### PR TITLE
feature/shell-commands

### PR DESCRIPTION
**Pull request by holyfabi**

### CODE REVIEW

File: src/Protector.php [hunk 1]
```diff
@@ -132,36 +134,41 @@ public function importDump(string $sourceFilePath, array $options): bool
...
+        $shellCommandImport = sprintf('mysql -h%s -u%s -p%s -D%s < %s 2> /dev/null',
+            escapeshellarg($this->connectionConfig['host']),
+            escapeshellarg($this->connectionConfig['username']),
+            escapeshellarg($this->connectionConfig['password']),
+            escapeshellarg($this->connectionConfig['database']),
+            escapeshellarg($filePath));
 
-            exec($shellCommandDropCreateDatabase);
-            exec($shellCommandImport);
+        $this->isExecEnabled();
```
[the remark this answers is not in the document]
Comment (holyfabi) [reply src/Protector.php]: Fixed.

File: src/Protector.php [hunk 2]
```diff
@@ -132,36 +134,41 @@ public function importDump(string $sourceFilePath, array $options): bool
         Storage::disk('local')->put($sourceFilePath, $this->getDisk()->get($sourceFilePath));
         $filePath = Storage::disk('local')->path($sourceFilePath);
 
-        try {
-            $shellCommandDropCreateDatabase = sprintf('mysql -h%s -u%s -p%s -e %s 2> /dev/null',
-                escapeshellarg($this->connectionConfig['host']),
-                escapeshellarg($this->connectionConfig['username']),
-                escapeshellarg($this->connectionConfig['password']),
-                escapeshellarg(sprintf('drop database %1$s; create database %1$s;', $this->connectionConfig['database'])));
+        $this->isExecEnabled();
```
Comment (jheusinger) [inline src/Protector.php:137]: Why would we generate all those shell calls, when it is impossible to execute them.

**Part 2**

File: src/Protector.php [hunk 3]
```diff
@@ -363,21 +370,30 @@ protected function generateDump(string $destinationFilePath, array $options = []
...
+                json_encode($options, JSON_UNESCAPED_UNICODE),
+                json_encode($this->getMetaData(), JSON_UNESCAPED_UNICODE));
 
-            return true;
+            $this->getDisk()->append($destinationFilePath, $metaData);
         } catch (Exception) {
-            return false;
+            throw new FailedDumpGenerationException('Error while creating the dump.');
         }
+
```
Comment (gael-connan-cybex) [inline src/Protector.php:396]: -

File: src/Protector.php (hunk 2 above)
Comment (holyfabi) [reply src/Protector.php]: Fixed.

File: src/Protector.php (hunk 3 above)
Comment (holyfabi) [reply src/Protector.php]: Done.

File: src/Protector.php [hunk 4]
```diff
@@ -363,20 +377,28 @@ protected function generateDump(string $destinationFilePath, array $options = []
 
         $this->createDirectory(Storage::disk('local')->path(dirname($destinationFilePath)));
 
-        try {
-            // Write dump using specific options.
-            exec(sprintf('mysqldump %s > %s 2> /dev/null',
+        $this->isExecEnabled();
```
Comment (jheusinger) [inline src/Protector.php:380]: Let's execute code and create a directory, even when we cant do anything with it.

+1

Comment (holyfabi) [reply src/Protector.php]: Fixed.

File: src/Protector.php [hunk 5]
```diff
@@ -771,6 +793,20 @@ protected function isSanctumActive(): bool
...
+    /**
+     * Throws an exception if Exec is deactivated.
+     *
+     * @throws ShellAccessDeniedException
+     */
+    public function isExecEnabled(): void
+    {
+        if (!function_exists('exec')) {
+            throw new ShellAccessDeniedException(
+              'Shell commands are disabled on your server, exec() must be enabled.'
```
Comment (gael-connan-cybex) [inline src/Protector.php:805]: <-

Comment (holyfabi) [reply src/Protector.php]: Done.

File: src/Exceptions/FailedShellCommandException.php [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+/**
+ * Class FailedShellCommandException
+ *
+ * Thrown if a shell command couldn't be executed properly.
+ *
+ * @package Cybex\Protector\Exceptions
+ */
+class FailedShellCommandException extends Exception
+{
+    //
```
[the remark this answers is not in the document]
Comment (holyfabi) [reply src/Exceptions/FailedShellCommandException.php]: Fixed.

**Review by jheusinger — APPROVED** [review]

**Pull request merged**